evmrpc/state.go

```diff
@@ -0,0 +1,85 @@
+package evmrpc
+
+import (
+	"context"
+	"encoding/hex"
+	"errors"
+	"fmt"
+	"strings"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/rpc"
+	"github.com/sei-protocol/sei-chain/x/evm/keeper"
+	rpcclient "github.com/tendermint/tendermint/rpc/client"
+)
+
+type StateAPI struct {
+	tmClient    rpcclient.Client
+	keeper      *keeper.Keeper
+	ctxProvider func() sdk.Context
+}
+
+func NewStateAPI(tmClient rpcclient.Client, k *keeper.Keeper, ctxProvider func() sdk.Context) *StateAPI {
+	return &StateAPI{tmClient: tmClient, keeper: k, ctxProvider: ctxProvider}
+}
+
+func (a *StateAPI) GetBalance(ctx context.Context, address common.Address, blockNr rpc.BlockNumber) (*uint64, error) {
+	block, err := getBlockNumber(ctx, a.tmClient, blockNr)
+	if err != nil {
+		return nil, err
+	}
+	if block != nil {
+		return nil, errors.New("block number not safe, finalized, latest, or pending")
+	}
+
+	balance := a.keeper.GetBalance(a.ctxProvider(), address)
+	return &balance, nil
+}
+
+func (a *StateAPI) GetCode(ctx context.Context, address common.Address, blockNr rpc.BlockNumber) ([]byte, error) {
+	block, err := getBlockNumber(ctx, a.tmClient, blockNr)
+	if err != nil {
+		return nil, err
+	}
+	if block != nil {
+		return nil, errors.New("block number not safe, finalized, latest, or pending")
+	}
+	code := a.keeper.GetCode(a.ctxProvider(), address)
+	return code, nil
+}
+
+func (a *StateAPI) GetStorageAt(ctx context.Context, address common.Address, hexKey string, blockNr rpc.BlockNumber) ([]byte, error) {
+	block, err := getBlockNumber(ctx, a.tmClient, blockNr)
+	if err != nil {
+		return nil, err
+	}
+	if block != nil {
+		return nil, errors.New("block number not safe, finalized, latest, or pending")
+	}
+	key, _, err := decodeHash(hexKey)
+	if err != nil {
+		return nil, fmt.Errorf("unable to decode storage key: %s", err)
+	}
+	state := a.keeper.GetState(a.ctxProvider(), address, key)
+	return state[:], nil
+}
+
+// decodeHash parses a hex-encoded 32-byte hash. The input may optionally
+// be prefixed by 0x and can have a byte length up to 32.
+func decodeHash(s string) (h common.Hash, inputLength int, err error) {
```

Comment:
is this the same as `common.HexToHash`, or there is some subtle difference?

Reply:
yea this was taken from go-ethereum here(https://github.com/ethereum/go-ethereum/blame/da55b23d21fb297fd01cfc1bebb13a3e8fe7f726/internal/ethapi/api.go#L750). It is quite similar, but the subtle difference is that decodeHash will return an error for invalid hex in storage keys whereas `HexToHash` will always resolve to some hash. The PR for that change in go-ethereum is here (https://github.com/ethereum/go-ethereum/commit/8ade5e6c144afb7ac389ae28b50cb68a0dbec7b8)